packages/protocol/test-sol/unit/common/UniswapFeeHandlerSeller.t.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.5.13;
+pragma experimental ABIEncoderV2;
+
+// Helper contracts
+import { Test } from "celo-foundry/Test.sol";
+import { TestConstants } from "@test-sol/constants.sol";
+
+import { UniswapFeeHandlerSeller } from "@celo-contracts/common/UniswapFeeHandlerSeller.sol";
+
+import { console2 as console } from "celo-foundry/Test.sol";
+
+contract UniswapFeeHandlerSellerTest is Test, TestConstants {
+  // Actors
+  address TOKEN_ADDRESS = actor("Arbitrary Token Address");
+  address ROUTER_ADDRESS = actor("Arbitrary Router Address");
+  address NON_OWNER_ADDRESS = actor("Arbitrary Non-Owner");
+  address ZERO_ADDRESS = address(0);
+  address UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS;
+
+  // Contract instance
+  UniswapFeeHandlerSeller uniswapFeeHandlerSeller;
+
+  function setUp() public {
+    uniswapFeeHandlerSeller = new UniswapFeeHandlerSeller(true);
+    UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS = uniswapFeeHandlerSeller.owner();
+  }
+}
+
+contract UniswapFeeHandlerSellerTest_SetRouter is UniswapFeeHandlerSellerTest {
+  function test_SetRouter_ShouldSucceedWhen_CalledByOwner() public {
+    vm.prank(UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS);
+    uniswapFeeHandlerSeller.setRouter(TOKEN_ADDRESS, ROUTER_ADDRESS);
+
+    assertEq(uniswapFeeHandlerSeller.getRoutersForToken(TOKEN_ADDRESS)[0], ROUTER_ADDRESS);
+  }
+
+  function test_SetRouter_ShouldRevertWhen_CalledByNonOwner() public {
+    vm.prank(NON_OWNER_ADDRESS);
+
+    vm.expectRevert("Ownable: caller is not the owner");
+    uniswapFeeHandlerSeller.setRouter(TOKEN_ADDRESS, ROUTER_ADDRESS);
+  }
+
+  function test_SetRouter_ShouldRevertWhen_SettingRouterToZeroAddress() public {
+    vm.prank(UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS);
+
+    vm.expectRevert("Router can't be address zero");
+    uniswapFeeHandlerSeller.setRouter(TOKEN_ADDRESS, ZERO_ADDRESS);
+  }
+
+  function test_SetRouter_ShouldRevertWhen_SettingMoreRoutesThanMaxAllowed() public {
+    uint256 MAX_NUMBER_ROUTERS_PER_TOKEN = 3; // This is a constant in the contract so it cannot be read from contract storage.
+
+    // Set exactly permitted number of routers
+    for (uint256 i = 0; i < 10; i++) {
+      /////// DEBUGGING //////
+      console.log("Setting router %d", i);
+      /////// DEBUGGING //////
+      vm.prank(UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS);
+      uniswapFeeHandlerSeller.setRouter(TOKEN_ADDRESS, ROUTER_ADDRESS);
+    }
+
+    vm.prank(UNISWAP_FEE_HANDLER_SELLER_OWNER_ADDRESS);
+    // vm.expectRevert("Max number of routers reached");
```

Comment:
I can't seem to get the contract to revert when the max number of routers are set per token. Although limit is hardcoded to 3, the `setRouter` function seems to accept more tokens than the limit.

```sh
Logs:
  Setting router 0
  Setting router 1
  Setting router 2
  Setting router 3
  Setting router 4
  Setting router 5
  Setting router 6
  Setting router 7
  Setting router 8
  Setting router 9
  Routers for token: 1
  ```

This `require` statement doesn't seem to be violated:

https://github.com/celo-org/celo-monorepo/blob/ed97cacf117c1d2a1b7c500ca7d4201c5c9e658e/packages/protocol/contracts/common/UniswapFeeHandlerSeller.sol#L156-L159

I wonder if it has to do with the way we use the `EnumerableSet` library

https://github.com/celo-org/celo-monorepo/blob/ed97cacf117c1d2a1b7c500ca7d4201c5c9e658e/packages/protocol/contracts/common/UniswapFeeHandlerSeller.sol#L30-L30

Reply:
Turns out I was just adding the same router address over and over. 

https://github.com/celo-org/celo-monorepo/blob/ed97cacf117c1d2a1b7c500ca7d4201c5c9e658e/packages/protocol/test-sol/unit/common/UniswapFeeHandlerSeller.t.sol#L56-L62

So it was not adding new routers, but overwriting the existing router. Managed to trigger the revert error by adding different addresses.

https://github.com/celo-org/celo-monorepo/blob/b1badbe7be809ba78b52b633e0d34a2054bef763/packages/protocol/test-sol/unit/common/UniswapFeeHandlerSeller.t.sol#L57-L61